source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1602 @@
+- Evaluating a Nix expression produces a single value.
+- Every *Nix file* (`.nix`) contains a single Nix expression.
+
+:::{note}
+To *evaluate* means to transform an expression according to the language rules until no further simplification is possible.
+:::
+
+All examples in this guide are valid Nix expressions.
+Each one is followed by the expected evaluation result.
+
+The following example is a Nix expression adding two numbers:
+
+```nix
+1 + 2
+```
+
+    3
+
+#### Interactive evaluation
+
+Use [`nix repl`][nix-repl] to evaluate Nix expressions interactively (by typing them on the command line):
+
+```console
+nix repl
+```
+
+    Welcome to Nix 2.5.1. Type :? for help.
+
+```console
+nix-repl> 1 + 2
+```
+
+    3
+
+[nix-repl]: https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-repl.html
+
+#### Evaluating Nix files
+
+Use [`nix-instantiate --eval`][nix-instantiate] to evaluate the expression in a Nix file.
+
+```console
+echo 1 + 2 > file.nix
+
+nix-instantiate --eval file.nix
+```
+
+    3
+
+<details><summary>Detailed explanation</summary>
+
+The first command writes `1 + 2` to a file `file.nix` in the current directory.
+The contents of `file.nix` are now `1 + 2`, which you can check with
+
+```console
+cat file.nix
+```
+
+    1 + 2
+
+The second command runs `nix-instantiate` with the `--eval` option on `file.nix`, which reads the file and evaluates the contained Nix expression.
+The resulting value is printed as output.
+
+`--eval` is required to evaluate the file and do nothing else.
+If `--eval` is omitted, `nix-instantiate` expects the expression in the given file to evaluate to a special data type called *Derivation*, which we will come back to at the end of this guide in [](build-tasks).
+
+If you do not need `file.nix` any more, remove it with
+
+```console
+rm file.nix
+```
+
+</details>
+
+:::{note}
+`nix-instantiate --eval` will evaluate `default.nix` if no file name is specified.
+
+```console
+echo 1 + 2 > default.nix
+
+nix-instantiate --eval
+```
+
+    3
+:::
+
+[nix-instantiate]: https://nixos.org/manual/nix/stable/command-ref/nix-instantiate.html
+
+## Reading the Nix language without fear
+
+You will quickly encounter Nix language expressions that may look very complicated.
+
+As with any programming language, the required amount of Nix language code closely matches the complexity of the problem it is supposed to solve, and reflects how well the problem – and its solution – is understood.
+
+Building software is a complex undertaking, and Nix both exposes and allows managing this complexity with the Nix language.
+
+The purpose of the Nix language is to create *build tasks*: precise descriptions of how contents of existing files are used to derive new files.
+
+:::{important}
+A build task in Nix is called *derivation*.
+:::
+
+The Nix language has only few basic constructs which can be combined arbitrarily:
+
+- primitive data types
+
+  as basic building blocks
```

Comment:
I guess by this you mean "Expressions representing their values directly", so e.g. `null`, `1`, `"str"`, `{ x = 10; }` but not `1 + 1`, `let x = 10; in x`, etc.

However, this category then also includes functions, e.g. `x: x` or `{ f = x: x; }`, which also evaluate to their values directly. But functions are in the next category already

Reply:
Yes, functions are also primitives, but as with the other things you pointed out this is purely for helping orient the reader by providing consistent structure in both overview and sections.

Reply:
> I guess by this you mean "Expressions representing their values directly"

Actually no, I'm not even thinking much in terms of expressions and values in this context, but instead want to address what I suppose are semantics: what do you *do* with these language objects?

In that sense (which is borrowed from Abelson & Sussman's SICP) functions are not primitives, even if the expressions declaring them do not evaluate further. They are means of combination. Bindings in function parameters are a bit more tricky, but intuitively it's not strictly a means of abstraction.

Maybe I'm relying on too ancient sources here, please enlighten me.

Reply:
I think this might be important from what you linked:

> In programming, we deal with two kinds of elements: functions and data. (Later we will discover that they are really not so distinct.)

Because in functional programming languages, functions are first-class citizens and can be passed around like any other data. This then makes the distinction between data and functions tricky, because functions can be both. In the end I think it's really just a matter of convention about how terms are defined, but this leads to a fuzzy categorization. But I get your idea of making this distinction for having a better tutorial flow.

Taking that into account though, I think it would be good to categorize it into "simple expressions" and "function declaration and application"